YAFCparser/Data/FactorioDataDeserializer_Context.cs

```diff
@@ -372,38 +371,55 @@ private Recipe CreateSpecialRecipe(FactorioObject production, string category, s
         private class DataBucket<TKey, TValue>: IEqualityComparer<List<TValue>>
         {
             private readonly Dictionary<TKey, IList<TValue>> storage = new Dictionary<TKey, IList<TValue>>();
-            private TValue[] def = Array.Empty<TValue>();
-            private bool seal;
-
-            // Replaces lists in storage with arrays. List with same contents get replaced with the same arrays
-            public void SealAndDeduplicate(TValue[] addExtra = null, IComparer<TValue> sort = null)
+            /// <summary>This function provides a default list of values for the key for when the key is not present in the storage.</summary>
+            /// <remarks>The provided function must *must not* return null.</remarks>
+            private Func<TKey,IEnumerable<TValue>> defaultList = NoExtraItems;
+
+            /// <summary>When true, it is not allowed to add new items to this bucket.</summary>
+            private bool isSealed;
+
+            /// <summary>
+            /// Replaces the list values in storage with array values while (optionally) adding extra values depending on the item.
+            /// </summary>
+            /// <remarks>
+            /// Lists with same contents get replaced with the same arrays for efficiency reasons.
+            /// </remarks>
+            /// <param name="addExtraItems">Function to provide extra items, *must not* return null.</param>
+            public void SealAndDeduplicate(Func<TKey,IEnumerable<TValue>> addExtraItems = null)
             {
-                if (addExtra != null)
-                    def = addExtra;
-                var mapDict = new Dictionary<List<TValue>, TValue[]>(this);
-                var vals = storage.ToArray();
-                foreach (var (key, value) in vals)
+                if (isSealed)
+                    throw new InvalidOperationException("Data bucket is already sealed");
+
+                if (addExtraItems != null)
+                    defaultList = addExtraItems;
+
+                Dictionary<List<TValue>, TValue[]> processedValues = new Dictionary<List<TValue>, TValue[]>(this);
+                KeyValuePair<TKey, IList<TValue>>[] values = storage.ToArray();
+                foreach ((TKey key, IList<TValue> value) in values)
                 {
-                    if (!(value is List<TValue> list)) 
+                    if (value is not List<TValue> list)
+                        // Unexpected type, (probably) never happens
                         continue;
-                    if (mapDict.TryGetValue(list, out var prev))
+                    if (processedValues.TryGetValue(list, out TValue[] prev))
```

Comment:
here one can clearly see how optimization is the root of all evil. it adds so much complexity to this class that it's really difficult to understand, and so easy to introduce bugs.

let's go one step back:
this dictionary is supposed to be a "simple" Dictionary<TKey, TValue[]>, but with the following performance concerns:

a) the values for the keys are added one after another and resizing arrays for every add is not wanted, therefore it's initially a Dictionary<TKey, List<TValue>> before it switches to TValue[] when there are no more values added.

b) there are likely to be multiple keys that have the same list of values, therefore when the adding of values has finished, all `List<TValue>` are processed and every group of equal lists is converted to a single shared `TValue[]`. This might be more than pure performance, if you do this kind of deduplication you can later check equality much easier by just comparing references. I doubt it however, because why would you need to compare compatible modules for different recipes? this class is generic however, didn't check whether it's used outside of the recipe/module association.

c) there are some `TValue` that are added to many (previously all) `TKey`, these are added implicitly as a "bulk operation"

d) there are likely to be lots of keys without any explicit data. for those no storage is necessary: non-existence of data means default values.

unfortunately with the default now being dynamic, most of these performance considerations need to be handled differently:

a) this is unchanged, but the switch of used types makes the whole class so much less readable

b) previously `processedValues.TryGetValue(list, ...)` was ok, because the same values were added, but now we actually have to calculate `completeList` first and use that as key for `processedValues`, otherwise the deduplication breaks the logic

c/d) `GetArray` now creates a new `TValue[]` every time without storing the calculated result (calling GetArray with same argument multiple times returns different results each time), but also without comparing them to existing lists (calling GetArray with different arguments returns different array instances with the same content). We should at least store the result of `defaultList(key).ToArray()` into `storage` for the next retrieval, but if we want to keep the same deduplication logic we actually have to keep `processedValues` around for the entire lifetime and use it here.

I have no idea why previously `GetArray` and `GetRaw` handled the default case differently (just returning def vs. storing def in the dictionary), but `GetRaw` has the same issues as `GetArray` regarding deduplication.

---

sorry, I don't want to sound harsh here, I really appreciate your effort, and this evaluation is purely based on reading the code, I haven't tested it at all. the thing is: just because it works for one modpack, or maybe even works without any flaws because of how this class is used, the class itself is a liability to keep around.

Reply:
No need to apologize! Thanks a lot for your full investigation and report! :smile: 

I tried to stay as close to the (working) original as possible. But I agree with you that this looks messy and having multiple use-cases in a single class...

I *think/feel* that `GetRaw()` is meant to be called before the data is 'sealed' and `GetArray()` for after. Hence their different approaches (and me not 'daring' to change it too much, also due to the lack of tests, so I wouldn't feel confident in doing so).

If I would have the job of cleaning this up I would
1. Add tests to make sure my cleanup wouldn't break anything
2. Split off the raw data and sealed data into separate classes
3. Maybe add different classes/abstractions for the different use-cases even to keep the simple stuff simple.
4. Only optimize if needed (as you mentioned it adds a lot of complexity and corner cases for bugs)

For this PR I would propose to keep the changes as close as possible, to prevent breaking unforeseen things. And add some future task/issue to clan this up. (maybe the whole analyser/solver, as it is very murky to me what it does and how).

Would you agree on this approach? (Assuming we are even willing to touch this code and clean it up)

Reply:
I could support a minimalistic approach: fix the logic mistakes, ignore the performance implications, no huge refactoring, no added unit tests - but only because I see a huge benefit in the change itself

Reply:
So basically what the current state of the change is? Do some minimal rework to support the dynamic extra items, while doing to minimal clean up of obsolete things as well and keeping the (broken/weird) current implementation.

Shall we merge the PR then?
And (optionally) create the issue to properly clean this up (if we feel it is required and benefiting the project)?

Reply:
fix what I described with my second b) paragraph (calculate `completeList` before lookup in `processedValues`), and merge it back, yes

I personally wouldn't create issues to clean up specific parts of the code, as long as everyone around here is basically a stranger. There are a few contributors, but I believe no one has the means and motivation to take over this project entirely (myself included). If that would change, the person in charge can decide upon areas to clean up.

Reply:
I had to study your comment b) again to see what you meant, but you are completely right that it might cause issues.

I removed the `processedList` completely, as it is meant as some (premature) optimization, but mainly adds complexity as I cannot see/notice any difference when loading the full pyanodons mod suite in YAFC (which is I guess the most extensive overhaul mod suite?)

Reply:
main concern was addressed. I'm closing this conversation now.